### DISCUSSION TITLE
Parallel Process Slice

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#3562] [Q&A] [answered]
Is it possible to improve the speed of `slice` with parallel processing? I've tried using `concurrent.futures`, both `ThreadPoolExecutor` and `ProcessPoolExecutor`, but the improvement only seems modest. 

Has anyone experimented with this, say, with the [Iso-Lines Contouring](https://docs.pyvista.org/examples/01-filter/contouring.html#iso-lines) example?

**Comment by adam-grant-hendry** [comment, marked as answer]
@akaszynski @banesullivan @adeak After experimenting a bit, looking at [Extract PolyLines from PolyData](https://kitware.github.io/vtk-examples/site/Python/PolyData/ExtractPolyLinesFromPolyData/), it seems infinitely faster to use `GenerateValues()` with `vtkCutter()` than explicitly looping in [`core/filters/data_set.py::slice_along_axis`](https://github.com/pyvista/pyvista/blob/7195d44217df23b1bd7640ff244cf9e0f73914fb/pyvista/core/filters/data_set.py#L733), for instance. What do you think?

```python
import numpy as np
import pyvista as pv

from pyvista import examples
from pyvista.core.filters import _get_output, _update_alg

from vtkmodules.vtkCommonDataModel import vtkPlane
from vtkmodules.vtkFiltersCore import vtkCutter

hills = examples.load_random_hills()

center = np.array(hills.center)
bounds = np.array(hills.bounds)

xmin, xmax, ymin, ymax, zmin, zmax = hills.bounds

n_slices = 50
progress_bar = True

plane = vtkPlane()

alg = vtkCutter()
alg.SetInputDataObject(hills)
alg.SetCutFunction(plane)
alg.GenerateValues(n_slices, zmin, zmax)

_update_alg(alg, progress_bar, 'Slicing')
contours = _get_output(alg)

plotter = pv.Plotter()
plotter.add_mesh(contours)
plotter.show()
```

![image](https://user-images.githubusercontent.com/59346180/200239383-ab3bddd8-53a1-467c-adcf-8d001465f55c.png)

**Reply by banesullivan** [reply]
Thanks for figuring this out and sharing! Would you be up for making a PR that implements this optimization?

**Comment by adam-grant-hendry** [comment]
@akaszynski @banesullivan @adeak Then, for an application like [Averaging Data on Individual Contours](https://github.com/pyvista/pyvista-support/issues/505), the problem becomes even simpler:

```python
import numpy as np
import pyvista as pv

from pyvista import examples
from pyvista.core.filters import _get_output, _update_alg
from tqdm import tqdm
from vtkmodules.vtkCommonDataModel import vtkPlane
from vtkmodules.vtkFiltersCore import vtkCutter

hills = examples.load_random_hills()

rrng = np.random.default_rng(seed=12345)

hills['Data'] = rrng.integers(  # Point data
    low=0,
	high=10,
	size=hills.points.shape[0],
)

hills['Averages'] = np.zeros_like(  # Point data
    hills['Data']
)

center = np.array(hills.center)
bounds = np.array(hills.bounds)

xmin, xmax, ymin, ymax, zmin, zmax = hills.bounds

n_slices = 100
progress_bar = True

plane = vtkPlane()

alg = vtkCutter()
alg.SetInputDataObject(hills)
alg.SetCutFunction(plane)
alg.GenerateValues(n_slices, zmin, zmax)

_update_alg(alg, progress_bar, 'Slicing')
contours = _get_output(alg)

contours = contours.cell_data_to_point_data()

connectivity = contours.connectivity()
labels = contours.point_arrays['RegionId']

for id_ in np.unique(labels):
    data = connectivity['Data'][labels == id_]
	connectivity['Averages'][labels == id_] = np.mean(data)

plotter = pv.Plotter()
plotter.add_mesh(connectivity, scalars='Averages')
plotter.show()
```

![image](https://user-images.githubusercontent.com/59346180/200243349-b6436063-bad3-4b7d-a0a5-3743e5184fb1.png)
